Backport west build --pristine to 1.14

carlescufi opened the pull request:
Fixes #16414

carlescufi committed:
west: build: Use officially documented CMake options

The -B option has always existed but was first officially documented in
CMake 3.13.0. In that same release the -S option was introduced,
replacing the old undocumented -H. Switch to using the officially
documented options.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
west: build: Add new pristine cmd-line and config option

Add a new command-line and build config option, `pristine`, that the
user can pass to `west build` or set in its configuration file(s) in
order to automatically trigger a pristine build on every build or
whenever west considers it required.

The option can take the following values:

- never: Never run the target
- always: Always run the pristine target before building
- auto: Run the pristine target when required

With `auto`, the pristine target will be run when running
west with an existing build folder containing a build system and:

- Selecting a different board from the one currently in the build system
- Selecting a different application from the one currently in the build
  system

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
scripts: west: Run pristine.cmake directly instead of the target

When making a build folder pristine until now we were running the
'pristine' build target. The issue with that is that ninja/make or
whatever build tool is being used might decide to re-run CMake itself if
some of the dependencies have changes. This might trigger an error that
is unfriendly and unnecessary, since the user is explicitly asking for
the build folder to be wiped before starting a fresh build.
To avoid this issue restor to running directly the CMake script that the
'pristine' build target itself uses, so as to make sure that the build
folder is wiped unconditionally regardless of changes made to the tree.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

mbolivar-nordic committed:
scripts: flake8 cleanup for west build

Clean up coding nits to match west's style.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

carlescufi commented on scripts/west_commands/build.py:
```diff
@@ -52,6 +53,10 @@
 
+class AlwaysIfMissing(argparse.Action):
+
+        def __call__(self, parser, namespace, values, option_string=None):
```
No, I do not thing it's worth the risk of changing the code more than it's needed

mbolivar replied:
I mean, it's whitespace changes, but fine 😉

carlescufi commented on doc/guides/west/config.rst:
```diff
@@ -165,14 +165,15 @@ extension commands (found in :file:`scripts/west_commands`).
-           target before building, if a build system is present and the build
-           will fail otherwise (e.g. the user has specified a different board or
-           application from the one previously used to make the build
+     - String. Controls the way in which ``west build`` may clean the build
```
Same

mbolivar replied:
It's documentation, the risk analysis is not the same and I disagree here.

dbkinder reviewed (approved):
doc changes look OK.

mbolivar reviewed (changes requested):
I think we should keep documentation looking the same for backports and mainline as much as possible.

carlescufi committed:
doc: west: Backport moving the build config opts doc

Backport to 1.14 the move of the configuration options that control the
west build command to the page where the command itself is described.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@mbolivar addressed the additional backport of linting and doc

mbolivar reviewed (approved):
LGTM, thanks!
